lib/pingurl.js

```diff
@@ -12,51 +12,54 @@ const TIMEOUT_MS = 5000;
 const TIMEOUT_WAIT_MS = 1000;
 const MAX_ATTEMPTS = 3;
 
+const sleep = ms => new Promise(r => setTimeout(r, ms));
+const getAgent = (url, a) => (a ? (url.match(/^https/) ? a.https : a.http) : undefined);
+
 /**
  * Attempt to GET a url, with retries/error handling
  */
-exports.ping = async (pingUrl, inputData, timeout, timeoutWait, agents) => {
-  let parsed = url.parse(pingUrl);
+exports.ping = async (url, inputData, timeout, timeoutWait, agents) => {
+  const opts = {
+    method: 'GET',
+    headers: exports.parseHeaders(inputData),
+    agent: getAgent(url, agents),
+  };
+  return exports.request(url, opts, null, timeout, timeoutWait);
+};
+
+/**
+ * Make a request with timeouts/retries
+ */
+exports.request = async (url, opts, data, timeout, timeoutWait) => {
+  timeout = timeout === undefined ? TIMEOUT_MS : timeout;
+  timeoutWait = timeoutWait === undefined ? TIMEOUT_WAIT_MS : timeoutWait;
+
+  // special cases
+  const parsed = parse(url);
   if (!parsed.host) {
-    throw new Error(`Invalid ping url: ${pingUrl}`);
+    throw new Error(`Invalid ping url: ${url}`);
   }
-  if (parsed.hostname === 'tps.doubleverify.com') {
+  if (parsed.host === 'tps.doubleverify.com') {
     throw new Error(`tps.doubleverify.com is not our friend`);
   }
-
-  // options
-  const opts = {
-    url: pingUrl,
-    host: parsed.hostname,
-    path: parsed.path,
-    port: parsed.port,
-    headers: exports.parseHeaders(inputData)
-  };
-  if (agents) {
-    opts.agents = agents;
-  }
-  timeout = (timeout === undefined) ? TIMEOUT_MS : timeout;
-  timeoutWait = (timeoutWait === undefined) ? TIMEOUT_WAIT_MS : timeoutWait;
-
-  // stop being sloooooooow
-  if (opts.host === 'sink.pdst.fm') {
+  if (parsed.host === 'sink.pdst.fm') {
     timeout = 1000;
```

Comment:
Maybe not for this PR, but could we also handle double verify with a reduced timeout instead of blocking entirely?
(I can't remember if their response was too slow or if it was the connection that was slow?)

Reply:
Do we still get `doubleverify` requests?  IIRC, I added this because it was failing in some way that didn't even trigger the request-timeout.  As if it was security-group blocked or something.